### PR TITLE
Change Btrieve cache to a bounded LRU cache

### PR DESCRIPTION
**Pull request by paladine**
Should be ready for review. @tuday2 maybe you can get get some metrics on memory comparisons with your large config?

### CODE REVIEW

**Comment by tuday2** [comment]
**31 modules**

**Before**
image

**After**
image

File: MBBSEmu/Configuration.cs [hunk 1]
```diff
@@ -32,19 +32,21 @@ public AppSettings()
...
+        public bool ModuleDoLoginRoutine => GetAppSettingsFromConfiguration<bool>("Module.DoLoginRoutine");
+        public bool TelnetEnabled => GetAppSettingsFromConfiguration<bool>("Telnet.Enabled");
+        public int TelnetPort => GetAppSettingsFromConfiguration<int>("Telnet.Port");
+        public bool TelnetHeartbeat => GetAppSettingsFromConfiguration<bool>("Telnet.Heartbeat");
+        public bool RloginEnabled => GetAppSettingsFromConfiguration<bool>("Rlogin.Enabled");
+        public int RloginPort => GetAppSettingsFromConfiguration<int>("Rlogin.Port");
         public string RloginRemoteIP => GetRemoteIPAppSettings("Rlogin.RemoteIP");
-        public bool RloginPortPerModule => GetAppSettings<bool>(ConfigurationRoot["Rlogin.PortPerModule"],"Rlogin.PortPerModule");
+        public bool RloginPortPerModule => GetAppSettingsFromConfiguration<bool>("Rlogin.PortPerModule");
         public string DatabaseFile => GetStringAppSettings("Database.File");
```
Comment (tuday2) [inline MBBSEmu/Configuration.cs:47]: Unnecessary blank line

Comment (tuday2) [reply MBBSEmu/Configuration.cs:47]: Also I added screenshots of the before/after to the PR. Everything seems to function fine.

File: MBBSEmu/Util/LRUCache.cs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+      var ret = _data.TryGetValue(key, out var v);
+      if (ret)
+      {
+        SetMostRecentlyUsed(v);
+
+        value = v._data;
+      }
+      else
+      {
+        value = default(TValue);
```
Comment (enusbaum) [inline MBBSEmu/Util/LRUCache.cs:173]: you should be able to simplify this to just `value = default`

File: MBBSEmu/Util/LRUCache.cs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+      }
+      return ret;
+    }
+
+    public IEnumerator<KeyValuePair<TKey,TValue>> GetEnumerator() => throw new NotSupportedException();
+
+    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => throw new NotSupportedException();
```
Comment (enusbaum) [inline MBBSEmu/Util/LRUCache.cs:180]: We can just import `System.Collections`

**Review by enusbaum — CHANGES_REQUESTED** [review]

File: MBBSEmu/Util/LRUCache.cs [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    /// </summary>
+    /// <value></value>
+    public int ListCount { get => _recentlyUsedList.Count; }
+
+    /// <summary>
+    ///   The most recently used key.
+    /// </summary>
+    public TKey MostRecentlyUsed { get => _recentlyUsedList.First.Value; }
+
+    public bool IsReadOnly { get => false; }
```
[the remark this answers is not in the document]
Comment (paladine) [reply MBBSEmu/Util/LRUCache.cs]: Required as part of `IDictionary`

**Review by enusbaum — APPROVED** [review]

**Pull request merged**